cmd/shell_test.go

```diff
@@ -228,3 +229,57 @@ func TestFailedConfigValidation(t *testing.T) {
 	test.AssertError(t, err, "Expected validation error")
 	test.AssertContains(t, err.Error(), "'required'")
 }
+
+func TestFailExit(t *testing.T) {
+	// Test that when Fail is called with a `defer AuditPanic()`,
+	// the program exits with a non-zero exit code and logs
+	// the result (but not stack trace).
+	// Inspired by https://go.dev/talks/2014/testing.slide#23
+	if os.Getenv("TIME_TO_DIE") == "1" {
+		defer AuditPanic()
+		Fail("tears in the rain")
+		return
+	}
+
+	// gosec points out that os.Args[0] is tainted, but we only run this as a test
+	// so we are not worried about it containing an untrusted value.
+	//nolint:gosec
+	cmd := exec.Command(os.Args[0], "-test.run=TestFailExit")
+	cmd.Env = append(os.Environ(), "TIME_TO_DIE=1")
+	output, err := cmd.CombinedOutput()
+	test.AssertError(t, err, "running a failing program")
+	test.AssertContains(t, string(output), "[AUDIT] tears in the rain")
+	// "goroutine" usually shows up in stack traces, so we check it
+	// to make sure we didn't print a stack trace.
+	test.AssertNotContains(t, string(output), "goroutine")
+}
+
+func testPanicStackTraceHelper() {
+	var x *int
+	*x = 1
+}
+
+func TestPanicStackTrace(t *testing.T) {
+	// Test that when a nil pointer dereference is hit after a
+	// `defer AuditPanic()`, the program exits with a non-zero
+	// exit code and prints the result (but not stack trace).
+	// Inspired by https://go.dev/talks/2014/testing.slide#23
+	if os.Getenv("AT_THE_DISCO") == "1" {
+		defer AuditPanic()
+		testPanicStackTraceHelper()
+		return
+	}
+
+	// gosec points out that os.Args[0] is tainted, but we only run this as a test
+	// so we are not worried about it containing an untrusted value.
+	//nolint:gosec
+	cmd := exec.Command(os.Args[0], "-test.run=TestPanicStackTrace")
+	cmd.Env = append(os.Environ(), "AT_THE_DISCO=1")
+	output, err := cmd.CombinedOutput()
+	test.AssertError(t, err, "running a failing program")
+	test.AssertContains(t, string(output), "nil pointer dereference")
+	// Note: we can confirm that we attempted to print a stack trace,
+	// though for some reason the actual lines of the stack trace don't
+	// appear to show up when calling `runtime.Stack()` in a unittest.
```

Comment:
Weird, I don't know why these would differ, given that `debug.Stack()` is documented to call `runtime.Stack()`.

Reply:
Well, the _old_ version as of last week passed a non-empty byte slice to runtime.Stack. Specifically it passed a slice with a backing array full of zeroes, a len of 8192 and a cap of 8192. We "fixed" that in https://github.com/letsencrypt/boulder/commit/e18507a43c58ab099d994b62f3a39effb529af89, but looking at that change now I see it's wrong. If `runtime.Stack()` was capable of growing a slice to the size it needs to hold the stack, it would take a `*[]byte`, not a `[]byte` (since it would need to be able to modify the caller's slice's backing array pointer, cap, and len). I was pretty sure I tested that PR manually and was getting stack traces, but in the absence of a unittest it's possible I got confused and did not manually test the final version.

Why is `runtime.Stack()` designed this way, rather than returning a newly allocated `[]byte` like `debug.Stack()` does? I'm not sure but I'm guessing it's out of a desire to avoid allocations while unwinding a panic, since allocations can fail and then you've got double-panic!

Reply:
Ah yep! I got one of these organically in my own integration tests just now:
```
23:08:17.484359 3 boulder-sa jMWPNwA [AUDIT] Panic caused by err: No ColumnMap in table crlShards type crlShardModel with field ID
23:08:17.484374 3 boulder-sa 6eet9AM [AUDIT] Stack Trace (Current frame) 
23:08:17.484384 4 boulder-sa uK-a7AY Stack Trace (All frames): 
```

So it seems like the solution here is to allocate a big buffer, or to use debug.Stacktrace to allocate the big buffer for us.

Reply:
Done! I wound up getting rid of the other-goroutine stack traces, which aren't typically useful.